__tests__/classes/ImageData.js

```diff
@@ -1,38 +1,133 @@
 describe('ImageData', () => {
-  it('should construct an ImageData object', () => {
-    const d = new ImageData(100, 200);
-    expect(d).toBeInstanceOf(ImageData);
+  it('should throw if less than 2 arguments are provided', () => {
+    expect(() => new ImageData()).toThrow(TypeError);
   });
 
-  it('should construct a Uint8ClampedArray', () => {
-    const d = new ImageData(100, 200);
-    expect(d.data).toBeInstanceOf(Uint8ClampedArray);
+  it('should throw if more than 3 arguments are provided', () => {
+    expect(() => new ImageData(new Uint8ClampedArray([0, 0, 0, 1]), 1, 1, 1))
+      .toThrow(TypeError);
   });
 
-  it('should construct a Uint8ClampedArray of proper size', () => {
-    const d = new ImageData(100, 200);
-    expect(d.data.length).toBe(100 * 200 * 4);
-  });
+  describe('new ImageData(width, height)', () => {
+    it('should construct an ImageData object', () => {
+      const d = new ImageData(100, 200);
+      expect(d).toBeInstanceOf(ImageData);
+    });
 
-  it('should set the width and height', () => {
-    const d = new ImageData(100, 200);
-    expect(d.width).toBe(100);
-    expect(d.height).toBe(200);
-  });
+    it('should construct a Uint8ClampedArray', () => {
+      const d = new ImageData(100, 200);
+      expect(d.data).toBeInstanceOf(Uint8ClampedArray);
+    });
 
-  it('should throw if width is not finite', () => {
-    expect(() => new ImageData(Infinity, 100)).toThrow(TypeError);
-  });
+    it('should construct a Uint8ClampedArray of proper size', () => {
+      const d = new ImageData(100, 200);
+      expect(d.data.length).toBe(100 * 200 * 4);
+    });
+
+    it('should set the width and height', () => {
+      const d = new ImageData(100, 200);
+      expect(d.width).toBe(100);
+      expect(d.height).toBe(200);
+    });
+
+    it('should throw if width is not finite', () => {
+      expect(() => new ImageData(Infinity, 100)).toThrow(RangeError);
+    });
 
-  it('should throw if width is 0', () => {
-    expect(() => new ImageData(0, 100)).toThrow(TypeError);
+    it('should throw if width is 0', () => {
+      expect(() => new ImageData(0, 100)).toThrow(RangeError);
+    });
+
+    it('should throw if height is not finite', () => {
+      expect(() => new ImageData(100, Infinity)).toThrow(RangeError);
+    });
+
+    it('should throw if height is 0', () => {
+      expect(() => new ImageData(100, 0)).toThrow(RangeError);
+    });
   });
 
-  it('should throw if height is not finite', () => {
-    expect(() => new ImageData(100, Infinity)).toThrow(TypeError);
+  describe('new ImageData(array, width)', () => {
+    it('should construct an ImageData object with the right infered height', () => {
```

Comment:
Just confirmed in the browser. This should be the only change we need.

```ts
      const data = new Uint8ClampedArray(800);
      const d = new ImageData(data, 200);
      expect(d).toBeInstanceOf(ImageData);
      expect(d.height).toBe(1);
      expect(d.data).toBe(data);
```

Reply:
Done :blush:

Reply:
I had mistakenly commented without looking at the whole test suite. Apparently a lot of the changes I was going to ask for were already in the test suite! Glad I found this one though. 

@hustcc Just another day of looking and we can pull this.